checker/cic.mli

```diff
@@ -170,6 +170,17 @@ type set_predicativity = ImpredicativeSet | PredicativeSet
 
 type engagement = set_predicativity
 
+(** {6 Conversion oracle} *)
+
+type level = Expand | Level of int | Opaque
+
+type oracle = {
+  var_opacity : level Id.Map.t;
+  cst_opacity : level Cmap.t;
+  var_trstate : Id.Pred.t;
```

Comment:
I understand that it was there before this PR, but do you know why we need 4 constructions? Couldn't the "transparent" state be expressed by the mapping being different from `Opaque`?

Reply:
It's from f308d977d616ab9a248a28993cca8e4b9cb49a46.

Reply:
Thanks, but that looks still mysterious to me. Why was it necessary to build these predicates instead of looking up in the maps?

Reply:
It seems in the beginning there were the 2 trstate values as global refs in conv_oracle. Then in 85a223796514d98211c06593d7ec72e56ed21e33 they were replaced by the 2 opacity values (as global refs) introducing the get_transp_state function (out of order comitting: the parent commit uses the new function). In dbd94e12c the refs become functional state and in f308d97 the trstate fields are added to the record type.

The predicates are given to some functions which take transparent_state, eg `Evarconv.solve_unif_constraints_with_heuristics` uses the transparent state from the oracle by default but is also used with `Typeclasses.classes_transparent_state ()`. To lookup in the maps those functions would have to take closures I guess.

Reply:
@maximedenes It's an invariant enforced by the API that the two last fields are logically deduced from the two first ones. Nonetheless, caching them in the data structure makes their computation O(1) when it was O(n) before, which seems it was a bottleneck.

Reply:
What is the predicate needed for?

Reply:
AFAICT it is used by Evarsolve, so it's probably a matter of unfolding in unification.

Reply:
I'm missing something, and it may be due to jetlag. Aren't clients of `var_trstate` going to lookup some constants in the predicate? Why not do it directly in `var_opacity`?

Reply:
Are you saying that some clients are iterating over `var_trstate`?

Reply:
No, the API used by Evarconv is using a transparent state, not a conv_oracle, see e.g. `Reductionops.infer_conv`. While the former can be logically deduced from the latter, it is more efficient to store it directly inside. We might use a conv_oracle instead but that would require updating the API.

Reply:
Yes, it seems to me like instead of caching it, we should have made unification directly take a `conv_oracle`. The code would be much simpler IIUC.

Reply:
Indeed, this should be the purpose of another PR. That said, it might still be more efficient to do it the way it is currently done. As far as I can trace it back, the end client of the transparent state is `CClosure.red_set`, which is a critical piece of code. Having a map which is bigger than needed because it contains strategies when the only needed info is opaqueness might end up being costly. So, we could try a branch to uniformize the API but it would definitely require serious benchmarking.